### PR TITLE
Add download Button and Matomo Integration to Sol2Uml plugin

### PR DESCRIPTION
**Pull request by joeizang**
Addresses #3310

### CODE REVIEW

File: libs/remix-ui/solidity-uml-gen/src/lib/components/UmlDownload.tsx [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+                  onClick={() => {
+                    _paq.push(['trackEvent', 'solidityumlgen', 'download', 'downloadAsPng'])
+                    props.download('png')
+                  }}
+                >
+                  <span
+                    id='umlPngDownloadBtn'
+                    data-id='umlPngDownload'
+                    onClick={() => {
+                      _paq.push(['trackEvent', 'solidityumlgen', 'download', 'downloadAsPng'])
```
Comment (Aniket-Engg) [inline libs/remix-ui/solidity-uml-gen/src/lib/components/UmlDownload.tsx:76]: same matomo event is added above already

File: libs/remix-ui/solidity-uml-gen/src/lib/components/UmlDownload.tsx [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+                  placement="left-start"
+                  tooltipId="solUmlgenDownloadAsPdfTooltip"
+                  tooltipClasses="text-nowrap"
+                  tooltipText={'Download UML diagram as a PDF file'}
+              >
+                <div
+                  data-id='umlPdfDownload'
+                  onClick={() => {
+                    _paq.push(['trackEvent', 'solUmlgen', 'download', 'downloadAsPdf'])
+                    props.download('pdf')
```
Comment (Aniket-Engg) [inline libs/remix-ui/solidity-uml-gen/src/lib/components/UmlDownload.tsx:100]: why there is an `onClick` on two elements?

File: libs/remix-ui/solidity-uml-gen/src/lib/solidity-uml-gen.tsx [hunk 3]
```diff
@@ -18,24 +21,34 @@ interface ActionButtonsProps {
...
+  }, [updatedSvg])
 
 
   const encoder = new TextEncoder()
   const data = encoder.encode(updatedSvg)
   const final = btoa(String.fromCharCode.apply(null, data))
 
+  const download = useCallback((fileType: UmlFileType) => {
+    if (umlCopy.length === 0) {
+      console.log('svg not valid yet!')
```
Comment (Aniket-Engg) [inline libs/remix-ui/solidity-uml-gen/src/lib/solidity-uml-gen.tsx:45]: remove log

File: libs/remix-ui/solidity-uml-gen/src/lib/solidity-uml-gen.tsx [hunk 4]
```diff
@@ -18,24 +21,34 @@ interface ActionButtonsProps {
...
+      return
+    }
+    console.log({ umlCopy })
```
Comment (Aniket-Engg) [inline libs/remix-ui/solidity-uml-gen/src/lib/solidity-uml-gen.tsx:48]: remove log

File: libs/remix-ui/solidity-uml-gen/src/lib/utilities/UmlDownloadStrategy.ts [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+      canvas.style.height = `${Math.round(img.naturalHeight/scale)}`.concat('px')
+      canvas.style.margin = '0'
+      canvas.style.padding = '0'
+      ctx.scale(window.devicePixelRatio, window.devicePixelRatio)
+      ctx.drawImage(img, 0, 0, Math.round(img.naturalWidth/scale), Math.round(img.naturalHeight/scale))
+      const png = canvas.toDataURL('image/png')
+      const a = document.createElement('a')
+      a.download = fileName.split('/')[1].split('.')[0].concat('.png')
+      a.href = png
+      console.log('about to click the link to download the png!!!')
```
Comment (Aniket-Engg) [inline libs/remix-ui/solidity-uml-gen/src/lib/utilities/UmlDownloadStrategy.ts:62]: remove log

**Review by Aniket-Engg — COMMENTED** [review]
Not sure why browser is showing that multiple files are being downloaded. Please check

File: libs/remix-ui/solidity-uml-gen/src/lib/components/UmlDownload.tsx (hunk 2 above)
Comment (joeizang) [reply libs/remix-ui/solidity-uml-gen/src/lib/components/UmlDownload.tsx:100]: must have a mistake on my part.

File: libs/remix-ui/solidity-uml-gen/src/lib/components/UmlDownload.tsx (hunk 1 above)
Comment (joeizang) [reply libs/remix-ui/solidity-uml-gen/src/lib/components/UmlDownload.tsx:76]: ah, thats true

File: libs/remix-ui/solidity-uml-gen/src/lib/solidity-uml-gen.tsx (hunk 3 above)
Comment (joeizang) [reply libs/remix-ui/solidity-uml-gen/src/lib/solidity-uml-gen.tsx:45]: yeah

File: libs/remix-ui/solidity-uml-gen/src/lib/solidity-uml-gen.tsx (hunk 4 above)
Comment (joeizang) [reply libs/remix-ui/solidity-uml-gen/src/lib/solidity-uml-gen.tsx:48]: sure

File: libs/remix-ui/solidity-uml-gen/src/lib/utilities/UmlDownloadStrategy.ts (hunk 5 above)
Comment (joeizang) [reply libs/remix-ui/solidity-uml-gen/src/lib/utilities/UmlDownloadStrategy.ts:62]: sure

File: libs/remix-ui/solidity-uml-gen/src/lib/components/UmlDownload.tsx [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+                placement="left-start"
+                tooltipId="solUmlgenDownloadAsPngTooltip"
+                tooltipClasses="text-nowrap"
+                tooltipText={'Download UML diagram as a PNG file'}
+            >
+                <div
+                  data-id='umlPngDownload'
+                  onClick={() => {
+                    _paq.push(['trackEvent', 'solidityumlgen', 'download', 'downloadAsPng'])
+                    props.download('png')
```
Comment (Aniket-Engg) [inline libs/remix-ui/solidity-uml-gen/src/lib/components/UmlDownload.tsx:69]: There is already an onClick on parent element

**Review by Aniket-Engg — COMMENTED** [review]
On trying to download, browser still shows that multiple files are being download, on not accepting it, it blocks further downloads

File: libs/remix-ui/solidity-uml-gen/src/lib/components/UmlDownload.tsx (hunk 6 above)
Comment (joeizang) [reply libs/remix-ui/solidity-uml-gen/src/lib/components/UmlDownload.tsx:69]: yeah

File: libs/remix-ui/solidity-uml-gen/src/lib/components/UmlDownload.tsx [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        </Dropdown.Toggle>
+        <Dropdown.Menu as={UmlCustomMenu} className="custom-dropdown-items">
+          <Dropdown.Item
+            onClick={() => {
+              _paq.push(['trackEvent', 'solidityumlgen', 'download', 'downloadAsPng'])
+              props.download('png')
+            }}
+          >
+            <CustomTooltip
+                placement="left-start"
```
Comment (LianaHus) [inline libs/remix-ui/solidity-uml-gen/src/lib/components/UmlDownload.tsx:65]: i think you have wrong indentation here

File: libs/remix-ui/solidity-uml-gen/src/lib/components/UmlDownload.tsx [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+                    data-id='umlPdfDownload'
+                    className='far fa-file-pdf pl-2'
+                  >
+                  </span>
+                  <span className="pl-2">
+                    Download as PDF
+                  </span>
+                </div>
+              </CustomTooltip>
+            </Dropdown.Item>
```
Comment (LianaHus) [inline libs/remix-ui/solidity-uml-gen/src/lib/components/UmlDownload.tsx:111]: here as well

File: libs/remix-ui/solidity-uml-gen/src/lib/components/UmlDownload.tsx [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+            <Dropdown.Divider />
+            <Dropdown.Item onClick={() => {
+                _paq.push(['trackEvent', 'solUmlgen', 'download', 'downloadAsPdf'])
+                props.download('pdf')
+              }}
+            >
+              <CustomTooltip
+                  placement="left-start"
+                  tooltipId="solUmlgenDownloadAsPdfTooltip"
```
Comment (LianaHus) [inline libs/remix-ui/solidity-uml-gen/src/lib/components/UmlDownload.tsx:93]: here

File: libs/remix-ui/solidity-uml-gen/src/lib/components/UmlDownload.tsx (hunk 7 above)
Comment (joeizang) [reply libs/remix-ui/solidity-uml-gen/src/lib/components/UmlDownload.tsx:65]: Right, fixing...!

File: libs/remix-ui/solidity-uml-gen/src/lib/components/UmlDownload.tsx (hunk 9 above)
Comment (joeizang) [reply libs/remix-ui/solidity-uml-gen/src/lib/components/UmlDownload.tsx:93]: right

File: libs/remix-ui/solidity-uml-gen/src/lib/components/UmlDownload.tsx (hunk 8 above)
Comment (joeizang) [reply libs/remix-ui/solidity-uml-gen/src/lib/components/UmlDownload.tsx:111]: yup!

File: libs/remix-ui/solidity-uml-gen/src/lib/utilities/UmlDownloadStrategy.ts [hunk 10]
```diff
@@ -0,0 +1,85 @@
+import jsPDF from 'jspdf'
+import 'svg2pdf.js'
```
Comment (bunsenstraat) [inline libs/remix-ui/solidity-uml-gen/src/lib/utilities/UmlDownloadStrategy.ts:2]: how is this used?

File: libs/remix-ui/solidity-uml-gen/src/lib/utilities/UmlDownloadStrategy.ts [hunk 11]
```diff
@@ -0,0 +1,85 @@
+import jsPDF from 'jspdf'
```
Comment (bunsenstraat) [inline libs/remix-ui/solidity-uml-gen/src/lib/utilities/UmlDownloadStrategy.ts:1]: can we lazy load it?

File: libs/remix-ui/solidity-uml-gen/src/lib/utilities/UmlDownloadStrategy.ts (hunk 10 above)
Comment (bunsenstraat) [reply libs/remix-ui/solidity-uml-gen/src/lib/utilities/UmlDownloadStrategy.ts:2]: svg2pdf I mean

Comment (joeizang) [reply libs/remix-ui/solidity-uml-gen/src/lib/utilities/UmlDownloadStrategy.ts:2]: No I am removing it now.

File: libs/remix-ui/solidity-uml-gen/src/lib/utilities/UmlDownloadStrategy.ts (hunk 11 above)
Comment (joeizang) [reply libs/remix-ui/solidity-uml-gen/src/lib/utilities/UmlDownloadStrategy.ts:1]: I think so.

**Review by bunsenstraat — APPROVED** [review]

**Pull request merged**